examples/simple-dg.py

```diff
@@ -360,72 +361,65 @@ def avg(self):
 
 def interior_trace_pair(discr, vec):
     i = discr.interp("vol", "int_faces", vec)
-    e = obj_array_vectorize(discr.opposite_face_connection(), i)
-    return TracePair("int_faces", i, e)
+    e = discr.opposite_face_connection()(i)
+    return TracePair("int_faces", interior=i, exterior=e)
 
 # }}}
 
 
 # {{{ wave equation bits
 
-def wave_flux(actx, discr, c, w_tpair):
-    u = w_tpair[0]
-    v = w_tpair[1:]
+def wave_flux(actx, discr, c, q_tpair):
+    u = q_tpair.u
+    v = q_tpair.v
 
-    normal = thaw(actx, discr.normal(w_tpair.where))
+    normal = thaw(actx, discr.normal(q_tpair.where))
 
-    flux_weak = flat_obj_array(
-            np.dot(v.avg, normal),
-            normal[0] * u.avg,
-            normal[1] * u.avg)
+    flux_weak = WaveState(
+            u=np.dot(v.avg, normal),
+            v=normal * u.avg)
 
     # upwind
     v_jump = np.dot(normal, v.ext-v.int)
-    flux_weak += flat_obj_array(
-            0.5*(u.ext-u.int),
-            0.5*normal[0]*v_jump,
-            0.5*normal[1]*v_jump,
-            )
+    flux_weak += WaveState(
+            u=0.5*(u.ext-u.int),
+            v=0.5*normal*v_jump)
 
-    flux_strong = flat_obj_array(
-            np.dot(v.int, normal),
-            u.int * normal[0],
-            u.int * normal[1],
+    flux_strong = WaveState(
+            u=np.dot(v.int, normal),
+            v=u.int * normal,
             ) - flux_weak
 
-    return discr.interp(w_tpair.where, "all_faces", c*flux_strong)
-
+    return discr.interp(q_tpair.where, "all_faces", c*flux_strong)
 
-def wave_operator(actx, discr, c, w):
-    u = w[0]
-    v = w[1:]
 
-    dir_u = discr.interp("vol", BTAG_ALL, u)
-    dir_v = discr.interp("vol", BTAG_ALL, v)
-    dir_bval = flat_obj_array(dir_u, dir_v)
-    dir_bc = flat_obj_array(-dir_u, dir_v)
+def wave_operator(actx, discr, c, q):
+    dir_q = discr.interp("vol", BTAG_ALL, q)
+    dir_bc = WaveState(u=-dir_q.u, v=dir_q.v)
 
     return (
-            - flat_obj_array(
-                -c*discr.div(v),
-                -c*discr.grad(u)
+            WaveState(
+                u=c*discr.div(q.v),
+                v=c*discr.grad(q.u)
                 )
-            +  # noqa: W504
+            -  # noqa: W504
             discr.inverse_mass(
                 discr.face_mass(
                     wave_flux(actx, discr, c=c,
-                        w_tpair=interior_trace_pair(discr, w))
+                        q_tpair=interior_trace_pair(discr, q))
                     + wave_flux(actx, discr, c=c,
-                        w_tpair=TracePair(BTAG_ALL, dir_bval, dir_bc))
+                        q_tpair=TracePair(BTAG_ALL, dir_q, dir_bc))
                     ))
                 )
 
-
 # }}}
 
 
 def rk4_step(y, t, h, f):
     k1 = f(t, y)
+    # FIXME: Test that these error if trying to add WaveState and 3-obj-array
+    #zz = y+k1
+    #zz = k1+y
```

Comment:
Needs doing.

Reply:
This should error out in `_binary_op`, right? It'll try to broadcast then see that the array containers don't have the same type (unless the object array contains `WaveState`s).

Reply:
Didn't until https://github.com/inducer/meshmode/pull/175/commits/727363e7f7ba6705919a0c0ca0f7ec9b6d4e4beb, which was mighty confusing.

Reply:
:facepalm: